### ISSUE TITLE
New file format for machine learning models

### ISSUE THREAD

**Issue by paulmueller** [ZELLMECHANIK-DRESDEN/dclab#78]
I propose a common file format for machine learning (ML) models (e.g. pre-trained neural nets for blood count). The idea is to train models using a simple Python script or with AIDeveloper and then load the model in dclab/Shape-Out/Shape-In for classification/filtering/sorting.

The file format should:
- be able to contain multiple models
- be human readable (except for the actual ML models)

I propose:
- a zip file containing model files and metadata files in the JSON format
- the .modc extension

A .modc file should contain:
- the models (e.g. protocol buffer .pb files from tensorflow)
  - when a model is applied to an event it outputs a number between 0 and 1 to indicate
    a probability of a classification match (1 is 100% probable)
  - [EDIT by @phidahl] it should also be possible to define one model that can classify an event. In that case, the `ml_score_???` features should sum up to a value `<=1` where the remainder could be considered uncertainty or so. This needs to be discussed and best practice in current ML applications should be taken into account.
  - [EDIT] The serialized model file should also contain the steps for pre- or post-processing. For tensorflow, this can
    be achieved as described in this tutorial: https://sayak.dev/tf.keras/preprocessing/2020/04/13/embedding-image-preprocessing-functions.html#Step-3:-SavedModel-plunge
- one meta data file per model with
  - a list of python libraries required to run the model (e.g. tensorflow) [possible security issues here]
  - a version number (e.g. a date when it was created for reproducible research)
  - sha256 hash and name of the model file
  - name of the ml_score feature(s) the model provides
  - the features used as input for the model
  - human readable names of the features (e.g. Red Blood Cells for `ml_score_rbc`)
- ~recipes to compute additional features/inputs [possible security issues here]~ -> should be covered by serialized preprocessing
- a readme file that describes what the model does

@phidahl Since our current analysis eco-system is Python-only, is this idea compatible with Shape-In (C++). We should probably stick to the tensorflow .pb file format. What are the restrictions?

**Comment by phidahl** [comment]
Hi Paul,

I agree with most of the points.

The results, however, could also be a vector of N numbers between 0..1 (sum should be 1.0 or < 1 ?) giving the probabilities that the given event belongs to one of these N classes. For blood count there will be more than 4 classes, and it’s more more straight forward to implement this classification in one model. We should think about how to handle uncertainty here:  What if an event does not belong to either of the given classes?

For ShapeIn the idea is to use the OpenCV implementation to read and apply the models since it is very fast and uses only CPU.
According to documentation it can read .pb or .pbtxt files (https://docs.opencv.org/4.3.0/d6/d0f/group__dnn.html#gac9b3890caab2f84790a17b306f36bd57 <https://docs.opencv.org/4.3.0/d6/d0f/group__dnn.html#gac9b3890caab2f84790a17b306f36bd57>) It’s been a while since I tested this, and this part of OpenCv is under active development so I’d have to repeat this with some reference .pb.

What I remember about Tensorflow is, that for training one is restricted to CPU or NVIDIA GPUs.. (Bad for mac computers) But there might be better OpenCL ways around this problem by now.

**Comment by paulmueller** [comment]
Thanks for the input. Yes, having one model with multiple outputs should also be supported. I will add it to the list. I think a normalization across classes is interesting - but it does not really allow a comparison across events, right?

**Comment by paulmueller** [comment]
I think we should not worry about training speed right now. This can be done on dedicated hardware or by people with a lot of time and AIDeveloper. We still have to figure out which format to promote: onnx, pytorch, tensorflow -- whichever one has the best support across Python and C++ libraries.

**Comment by paulmueller** [comment]
@maikherbig Thanks for the insight. I guess it's gonna be .pb then. BTW we are not reinventing the wheel. We only need additional metadata about the model and the features used, and possibly additional Python scripts (as listed above). Even if the .pb file format supports all of these things, I would also like to hash the .pb file to verify it's the correct version. JSON files are human-readable (with proper line breaks and indentation of course).

**Comment by maikherbig** [comment]
Few more thoughts.
## json vs xlsx
To get a model, you train it sometimes for thousands of iterations. Each time, a few metrics are saved (accuracy, f1,..). Hence, the result is a table. That information belongs to the meta-information of the model. I find it convenient to save (small) tables as excel file as everybody can open and work with them immediately. Is there a particular advantage of JSON for such relatively small tables?

## TensorFlow vs OpenCV' dnn module
I suppose you are still trying to minimize the size of ShapeOut. Therefore, I would like to point out that TensorFlow has actually a footprint of approx. 460MB. 
Maybe a more memory efficient solution would be to go for OpenCV's dnn module, which was already suggested to be used in ShapeIn. Using the same tool in ShapeIn and ShapeOut might also make it easier to synchronize.

**Comment by paulmueller** [comment]
JSON is actually not really good for tables. Here, I would prefer tsv. I agree that the evolution of the model is important information (and so would be the training data - are you storing hashes of the raw training data?). I would suggest a separate folder "supplements" in the zip file for such information. Since there will not be strict rules for supplements, I think xls is fine there.

OpenCV would indeed be a viable option, because nowadays they have wheels for all platforms.

**Comment by maikherbig** [comment]
Yes, the meta-files of AID also contain the hash (have a look into the .zip I sent earlier).
This is very useful when loading such a meta file to restore a session in AID. If the path to the files has changed, the hash is used to find the files back. Loading such a session restores the table of data used as shown below. This means information is loaded which files are used and also how many images from which file per training iteration.
![LoadSession](https://user-images.githubusercontent.com/14074051/83963889-5bb8d280-a8a9-11ea-9e50-162bff760730.png)

**Comment by maikherbig** [comment]
@alfrem just told me about this new tool:
https://github.com/uber/neuropod
Neuropod allows to run inference for many model-formats. Neuropod provides tools to easily convert keras models to the neuropod format. This neuropod format can then be used for inference identically on Python and C++! Maybe it would make sense to compare the inference times (pure tensorflow, OpenCV-dnn, neuropod and maybe also [keras2cpp](https://github.com/pplonski/keras2cpp)).

**Comment by paulmueller** [comment]
Exporting models from tensorflow and loading them with OpenCV is not trivial.

- The model has to be exported as a frozen graph (e.g. https://github.com/opencv/opencv/issues/16582#issuecomment-603819498, https://github.com/opencv/opencv/issues/16879#issuecomment-603815872).
- This only seems to work with predefined models and not with custom models (https://stackoverflow.com/questions/60826380/opencv-cant-create-layer-map-shape-of-type-shape-in-function-getlayerinst), because a .pbtxt file is required (https://jeanvitor.com/tensorflow-object-detecion-opencv/)

TODO:

- Further reading: https://medium.com/@sathualab/how-to-use-tensorflow-graph-with-opencv-dnn-module-3bbeeb4920c5
- Find out how hard it is to generate .pbtxt files (e.g. https://github.com/opencv/opencv/blob/master/samples/dnn/tf_text_graph_ssd.py)

**Comment by maikherbig** [comment]
### Good news
it works with OpenCV 4.4.0.42!
I used to have "opencv-contrib-python-headless 4.1.1.26" and got all sorts of errors when trying to load and use a frozen .pb file (generated using the model conversion tool in AIDeveloper). 
After an update using: "pip install opencv-python-headless==4.4.0.42", everything suddenly worked perfectly!
I wrote a script (aid_OpenCV_dnn.py) that contains a testing function (test_opencv_dnn), which tests if the original model (.model) and the frozen model (.pb) return the same predictions. The attached .zip contains all required files:
[OpenCV_dnn_test.zip](https://github.com/ZELLMECHANIK-DRESDEN/dclab/files/5404789/OpenCV_dnn_test.zip)
 
When having a look at the testing function you will see how to:
- find the input image size, required by the model
- find the image normalization function that was used when the model was trained (same normalization needs to be applied when using the model)
- apply cropping and normalization method to rtdc images
- load a model (.pb)
- forward an image through the model to get predictions

I also tested another model (which has a dropout layer) -> no issues :)


the attached zip also contains aid_bin, aid_img, and aid_start which are scripts from AIDeveloper (https://github.com/maikherbig/AIDeveloper/tree/master/AIDeveloper)

### How to convert (freeze) a model in AIDeveloper?
- Open AID
- Go to the History tab
- Click "Load Model" (button on the left)
- Choose target format ("Frozen TensorFlow .pb" or "Optimized TensorFlow .pb") in drop-down menu on the right
- Click "Convert" (button on the right)

**Comment by paulmueller** [comment]
I am closing this one, because the basic .modc model file support is now implemented in dclab 0.30.0. There is a write-up of all the new functionalities in the new ML section of the documentation: https://dclab.readthedocs.io/en/stable/sec_av_ml.html

For all the other issues that were raised during this conversation, please see:
- #82 
- #83 
- #84
